backend_new/src/services/application-csv-export.service.ts

```diff
@@ -0,0 +1,653 @@
+    const filename = join(
+      process.cwd(),
+      `src/temp/listing-${
+        queryParams.listingId
+      }-applications-${new Date().getTime()}.csv`,
+    );
+    await this.createCsv(filename, queryParams);
+    const file = createReadStream(filename);
+    file.on('end', () => {
+      fs.unlink(filename, (err) => {
+        if (err) {
+          console.error(`Error deleting ${filename}`);
+          throw err;
+        }
+      });
+    });
+    return new StreamableFile(file);
+  }
+
+  /**
+   *
+   * @param filename
+   * @param queryParams
+   * @returns a promise with SuccessDTO
+   */
+  async createCsv(
+    filename: string,
+    queryParams: ApplicationCsvQueryParams,
+  ): Promise<SuccessDTO> {
+    if (queryParams.includeDemographics) {
+      view.csv.demographics = true;
+    }
+
+    const applications = await this.prisma.applications.findMany({
+      include: view.csv,
+      where: {
+        listingId: queryParams.listingId,
+      },
+    });
+
+    // get all multiselect questions for a listing to build csv headers
+    const multiSelectQuestions =
+      await this.multiselectQuestionService.findByListingId(
+        queryParams.listingId,
+      );
+
+    // get maxHouseholdMembers
+    const maxHouseholdMembers = await this.maxHouseholdMembers();
+
+    const csvHeaders = await this.getCsvHeaders(
+      maxHouseholdMembers,
+      multiSelectQuestions,
+      queryParams.includeDemographics,
+    );
+
+    return new Promise((resolve, reject) => {
+      // create stream
+      const writableStream = fs.createWriteStream(`${filename}`);
+      writableStream
+        .on('error', (err) => {
+          console.log('csv writestream error');
+          console.log(err);
+          reject(err);
+        })
+        .on('finish', () => {
+          console.log('finished');
+        })
+        .on('close', () => {
+          console.log('stream closed');
+          resolve({
+            success: true,
+          });
+        })
+        .on('open', () => {
+          writableStream.write(
+            csvHeaders.map((header) => header.label).join(',') + '\n',
+          );
+
+          // now loop over applications and write them to file
+          applications.forEach((app) => {
+            let row = '';
+            let preferences: ApplicationMultiselectQuestion[];
+            csvHeaders.forEach((header, index) => {
+              let multiselectQuestionValue = false;
+              let parsePreference = false;
+              let value = header.path.split('.').reduce((acc, curr) => {
+                // return preference/program as value for the format function to accept
+                if (multiselectQuestionValue) {
+                  return acc;
+                }
+
+                if (parsePreference) {
+                  // curr should equal the preference id we're pulling from
+                  if (!preferences) {
+                    preferences = JSON.parse(app.preferences as string);
```

Comment:
The saved preferences are not saved as strings so this JSON.parse fails. with any non-seeded applications. See comment about the seeded data